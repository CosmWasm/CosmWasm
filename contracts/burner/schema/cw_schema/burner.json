{
  "contract_name": "burner",
  "contract_version": "0.0.0",
  "idl_version": "1.0.0",
  "instantiate": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "InstantiateMsg",
        "description": "A placeholder where we don't take any input",
        "type": "struct",
        "properties": {}
      }
    ]
  },
  "execute": null,
  "query": null,
  "migrate": {
    "type": "v1",
    "root": 2,
    "definitions": [
      {
        "name": "String",
        "type": "string"
      },
      {
        "name": "u32",
        "type": "integer",
        "precision": 32,
        "signed": false
      },
      {
        "name": "MigrateMsg",
        "type": "struct",
        "properties": {
          "delete": {
            "description": "Optional amount of items to delete in this call.\nIf it is not provided, nothing will be deleted.\nYou can delete further items in a subsequent execute call.",
            "value": 1
          },
          "payout": {
            "description": "The address we send all remaining balance to",
            "value": 0
          }
        }
      }
    ]
  },
  "sudo": null,
  "responses": null
}
