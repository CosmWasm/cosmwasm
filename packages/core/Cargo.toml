[package]
name = "cosmwasm-core"
version = { workspace = true }
edition = "2021"
description = "Components of cosmwasm-std that can be used in no_std environments"
repository = "https://github.com/CosmWasm/cosmwasm/tree/main/packages/core"
license = "Apache-2.0"
readme = "README.md"
