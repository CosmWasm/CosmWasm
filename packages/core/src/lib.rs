//! cosmwasm-core contains components of cosmwasm-std that can be used in a no_std environment.
//! All symbols are re-exported by cosmwasm-std, such that contract developers don't need to
//! add this dependency directly. It is recommended to only use cosmwasm-std whenever possible.

#![cfg_attr(not(feature = "std"), no_std)]

#[macro_use]
extern crate alloc;

#[cfg(test)]
extern crate std; // allow for file I/O during tests

mod addresses;
mod binary;
mod encoding;
mod errors;
mod forward_ref;
mod hex_binary;
mod math;
mod timestamp;

#[doc(hidden)]
pub mod __internal;

use crate::__internal__forward_ref_partial_eq as forward_ref_partial_eq;

pub use crate::addresses::{instantiate2_address, Addr, CanonicalAddr, Instantiate2AddressError};
pub use crate::binary::Binary;
pub use crate::encoding::{from_base64, from_hex, to_base64, to_hex};
pub use crate::errors::{
    AggregationError, AggregationPairingEqualityError, CheckedFromRatioError,
    CheckedMultiplyFractionError, CheckedMultiplyRatioError, CoinFromStrError, CoinsError,
    ConversionOverflowError, CoreError, CoreResult, DivideByZeroError, DivisionError,
    OverflowError, OverflowOperation, RecoverPubkeyError, RoundDownOverflowError,
    RoundUpOverflowError, SystemError, VerificationError,
};
pub use crate::hex_binary::HexBinary;
pub use crate::math::{
    Decimal, Decimal256, Decimal256RangeExceeded, DecimalRangeExceeded, Fraction, Int128, Int256,
    Int512, Int64, Isqrt, SignedDecimal, SignedDecimal256, SignedDecimal256RangeExceeded,
    SignedDecimalRangeExceeded, Uint128, Uint256, Uint512, Uint64,
};
pub use crate::timestamp::Timestamp;

/// Exposed for testing only
/// Both unit tests and integration tests are compiled to native code, so everything in here does not need to compile to Wasm.
#[cfg(not(target_arch = "wasm32"))]
pub mod testing;
