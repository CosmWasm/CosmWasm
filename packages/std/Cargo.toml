[package]
name = "cosmwasm-std"
version = "1.2.5"
authors = ["Ethan Frey <ethanfrey@users.noreply.github.com>"]
edition = "2021"
description = "Standard library for Wasm based smart contracts on Cosmos blockchains"
repository = "https://github.com/CosmWasm/cosmwasm/tree/main/packages/std"
license = "Apache-2.0"
readme = "README.md"

[package.metadata.docs.rs]
features = ["abort", "stargate", "staking", "ibc3", "cosmwasm_1_2"]

[features]
default = ["iterator", "abort", "std"]
std = ["base64/std", "uint/std", "serde-json-wasm/std", "forward_ref", "schemars", "thiserror/std"]
abort = []
# iterator allows us to iterate over all DB items in a given range
# optional as some merkle stores (like tries) don't support this
# given Ethereum 1.0, 2.0, Substrate, and other major projects use Tries
# we keep this optional, to allow possible future integration (or different Cosmos Backends)
iterator = []
# staking exposes bindings to a required staking moudle in the runtime, via new
# CosmosMsg types, and new QueryRequest types. This should only be enabled on contracts
# that require these types, so other contracts can be used on systems with eg. PoA consensus
staking = []
# backtraces provides much better context at runtime errors (in non-wasm code)
# at the cost of a bit of code size and performance.
# This feature requires Rust nightly because it depends on the unstable backtrace feature.
backtraces = []
# stargate enables stargate-dependent messages and queries, like raw protobuf messages
# as well as ibc-related functionality
stargate = []
# ibc3 extends ibc messages with ibc-v3 only features. This should only be enabled on contracts
# that require these types. Without this, they get the smaller ibc-v1 API.
ibc3 = ["stargate"]
# This feature makes `BankQuery::Supply` available for the contract to call, but requires
# the host blockchain to run CosmWasm `1.1.0` or higher.
cosmwasm_1_1 = []
# This feature makes `GovMsg::VoteWeighted` available for the contract to call, but requires
# the host blockchain to run CosmWasm `1.2.0` or higher.
cosmwasm_1_2 = ["cosmwasm_1_1"]


[dependencies]
base64 =  {version = "0.13.0", default-features = false, features = ["alloc"]}
cosmwasm-derive = { path = "../derive", version = "1.2.5" }
derivative = { version = "2", default-features = false, features = ["use_core"] }
hex = { version = "0.4", default-features = false, features = ["alloc"] }
schemars = {version = "0.8.3", optional = true}
sha2 = { version = "0.10.3", default-features = false}
serde = { version = "1.0.103", default-features = false, features = ["derive", "alloc"] }
serde-json-wasm = { git = "https://github.com/dzmitry-lahoda-forks/serde-json-wasm", rev = "8a7e522c0e4de36a6dfb535766f26a9941017d81", default-features = false}
thiserror = { version = "1.0.38", package = "thiserror-core", default-features = false }

uint = { version = "0.9.3", default-features = false }

forward_ref = { version = "1", default-features = false, optional = true}

[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
cosmwasm-crypto = { path = "../crypto", version = "1.2.5", optional = true}


[dev-dependencies]
cosmwasm-schema = { path = "../schema" }
# The chrono dependency is only used in an example, which Rust compiles for us. If this causes trouble, remove it.
chrono = { version = "0.4", default-features = false, features = ["alloc", "std"] }
hex-literal = "0.3.1"
serde_json = "1.0.81"
