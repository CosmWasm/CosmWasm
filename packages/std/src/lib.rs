#[macro_use]
extern crate alloc;

// Exposed on all platforms

mod assertions;
mod checksum;
mod coin;
mod coins;
mod conversion;
mod deps;
mod ibc;
mod import_helpers;
#[cfg(feature = "iterator")]
mod iterator;
mod metadata;
mod never;
mod pagination;
mod panic;
mod query;
mod results;
mod sections;
mod serde;
mod stdack;
mod storage;
mod traits;
mod types;
mod uuid;

/// This module is to simplify no_std imports
pub(crate) mod prelude;

/// This modules is very advanced and will not be used directly by the vast majority of users.
/// We want to offer it to ensure a stable storage key composition system but don't encourage
/// contract devs to use it directly.
pub mod storage_keys;

pub use crate::checksum::{Checksum, ChecksumError};
pub use crate::coin::{coin, coins, has_coins, Coin};
pub use crate::coins::Coins;
pub use crate::deps::{Deps, DepsMut, OwnedDeps};
pub use crate::ibc::IbcChannelOpenResponse;
pub use crate::ibc::{
    Ibc3ChannelOpenResponse, IbcAcknowledgement, IbcBasicResponse, IbcChannel, IbcChannelCloseMsg,
    IbcChannelConnectMsg, IbcChannelOpenMsg, IbcEndpoint, IbcMsg, IbcOrder, IbcPacket,
    IbcPacketAckMsg, IbcPacketReceiveMsg, IbcPacketTimeoutMsg, IbcReceiveResponse, IbcTimeout,
    IbcTimeoutBlock,
};
#[cfg(feature = "iterator")]
pub use crate::iterator::{Order, Record};
pub use crate::metadata::{DenomMetadata, DenomUnit};
pub use crate::never::Never;
pub use crate::pagination::PageRequest;
pub use crate::query::{
    AllBalanceResponse, AllDelegationsResponse, AllDenomMetadataResponse, AllValidatorsResponse,
    BalanceResponse, BankQuery, BondedDenomResponse, ChannelResponse, CodeInfoResponse,
    ContractInfoResponse, CustomQuery, DecCoin, Delegation, DelegationResponse,
    DelegationRewardsResponse, DelegationTotalRewardsResponse, DelegatorReward,
    DelegatorValidatorsResponse, DelegatorWithdrawAddressResponse, DenomMetadataResponse,
    DistributionQuery, FullDelegation, GrpcQuery, IbcQuery, ListChannelsResponse, PortIdResponse,
    QueryRequest, StakingQuery, SupplyResponse, Validator, ValidatorResponse, WasmQuery,
};
#[cfg(all(feature = "stargate", feature = "cosmwasm_1_2"))]
pub use crate::results::WeightedVoteOption;
pub use crate::results::{
    attr, wasm_execute, wasm_instantiate, AnyMsg, Attribute, BankMsg, ContractResult, CosmosMsg,
    CustomMsg, Empty, Event, MsgResponse, QueryResponse, Reply, ReplyOn, Response, SubMsg,
    SubMsgResponse, SubMsgResult, SystemResult, WasmMsg,
};
#[cfg(feature = "staking")]
pub use crate::results::{DistributionMsg, StakingMsg};
#[cfg(feature = "stargate")]
pub use crate::results::{GovMsg, VoteOption};
#[allow(deprecated)]
pub use crate::serde::{
    from_binary, from_json, from_slice, to_binary, to_json_binary, to_json_string, to_json_vec,
    to_vec,
};
pub use crate::stdack::StdAck;
pub use crate::storage::MemoryStorage;
pub use crate::traits::{Api, Querier, QuerierResult, QuerierWrapper, Storage};
pub use crate::types::{BlockInfo, ContractInfo, Env, MessageInfo, TransactionInfo};
pub use crate::uuid::{new_uuid, Uuid};

// Exposed in wasm build only

#[cfg(target_arch = "wasm32")]
mod exports;
#[cfg(target_arch = "wasm32")]
mod imports;
#[cfg(target_arch = "wasm32")]
mod memory; // Used by exports and imports only. This assumes pointers are 32 bit long, which makes it untestable on dev machines.

#[cfg(target_arch = "wasm32")]
pub use crate::exports::{do_execute, do_instantiate, do_migrate, do_query, do_reply, do_sudo};
#[cfg(all(feature = "stargate", target_arch = "wasm32"))]
pub use crate::exports::{
    do_ibc_channel_close, do_ibc_channel_connect, do_ibc_channel_open, do_ibc_packet_ack,
    do_ibc_packet_receive, do_ibc_packet_timeout,
};
#[cfg(target_arch = "wasm32")]
pub use crate::imports::{ExternalApi, ExternalQuerier, ExternalStorage};

/// Exposed for testing only
/// Both unit tests and integration tests are compiled to native code, so everything in here does not need to compile to Wasm.
#[cfg(not(target_arch = "wasm32"))]
pub mod testing;

// Re-exports

pub use cosmwasm_core::CoreError as StdError;
pub use cosmwasm_core::CoreResult as StdResult;
pub use cosmwasm_core::{
    from_base64, from_hex, instantiate2_address, to_base64, to_hex, Addr, Binary, CanonicalAddr,
    CheckedFromRatioError, CheckedMultiplyFractionError, CheckedMultiplyRatioError,
    CoinFromStrError, CoinsError, ConversionOverflowError, Decimal, Decimal256,
    Decimal256RangeExceeded, DecimalRangeExceeded, DivideByZeroError, DivisionError, Fraction,
    HexBinary, Instantiate2AddressError, Int128, Int256, Int512, Int64, Isqrt, OverflowError,
    OverflowOperation, RecoverPubkeyError, SignedDecimal, SignedDecimal256,
    SignedDecimal256RangeExceeded, SignedDecimalRangeExceeded, SystemError, Timestamp, Uint128,
    Uint256, Uint512, Uint64, VerificationError,
};

#[cfg(not(target_arch = "wasm32"))]
pub use cosmwasm_core::assert_approx_eq;

pub use cosmwasm_derive::entry_point;
