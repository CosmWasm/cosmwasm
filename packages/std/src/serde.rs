// This file simply re-exports some methods from serde_json
// The reason is two fold:
// 1. To easily ensure that all calling libraries use the same version (minimize code size)
// 2. To allow us to switch out to eg. serde-json-core more easily

use core::any::type_name;
use serde::{de::DeserializeOwned, Serialize};

use crate::Binary;
use crate::{StdError, StdResult};

/// Deserializes the given JSON bytes to a data structure.
///
/// Errors if the input is not valid JSON or cannot be deserialized to the given type.
pub fn from_json<T: DeserializeOwned>(value: impl AsRef<[u8]>) -> StdResult<T> {
    serde_json::from_slice(value.as_ref()).map_err(|e| StdError::parse_err(type_name::<T>(), e))
}

/// Serializes the given data structure as a JSON byte vector.
pub fn to_json_vec<T>(data: &T) -> StdResult<Vec<u8>>
where
    T: Serialize + ?Sized,
{
    serde_json::to_vec(data).map_err(|e| StdError::serialize_err(type_name::<T>(), e))
}

/// Serializes the given data structure as a JSON string.
pub fn to_json_string<T>(data: &T) -> StdResult<String>
where
    T: Serialize + ?Sized,
{
    serde_json::to_string(data).map_err(|e| StdError::serialize_err(type_name::<T>(), e))
}

/// Serializes the given data structure as JSON bytes.
pub fn to_json_binary<T>(data: &T) -> StdResult<Binary>
where
    T: Serialize + ?Sized,
{
    to_json_vec(data).map(Binary::new)
}

#[cfg(test)]
mod tests {
    use super::*;
    use core::num::{
        NonZeroI128, NonZeroI16, NonZeroI32, NonZeroI64, NonZeroI8, NonZeroU128, NonZeroU16,
        NonZeroU32, NonZeroU64, NonZeroU8,
    };
    use proptest::{prop_assert_eq, property_test};
    use serde::Deserialize;

    use crate::msgpack::{from_msgpack, to_msgpack_vec};

    #[derive(Serialize, Deserialize, Debug, PartialEq)]
    #[serde(rename_all = "snake_case")]
    enum SomeMsg {
        Refund {},
        ReleaseAll {
            image: String,
            amount: u32,
            time: u64,
            karma: i32,
        },
        Cowsay {
            text: String,
        },
    }

    #[test]
    fn to_json_vec_works() {
        let msg = SomeMsg::Refund {};
        let serialized = to_json_vec(&msg).unwrap();
        assert_eq!(serialized, br#"{"refund":{}}"#);

        let msg = SomeMsg::ReleaseAll {
            image: "foo".to_string(),
            amount: 42,
            time: 9007199254740999, // Number.MAX_SAFE_INTEGER + 7
            karma: -17,
        };
        let serialized = String::from_utf8(to_json_vec(&msg).unwrap()).unwrap();
        assert_eq!(
            serialized,
            r#"{"release_all":{"image":"foo","amount":42,"time":9007199254740999,"karma":-17}}"#
        );
    }

    #[test]
    fn from_json_works() {
        let deserialized: SomeMsg = from_json(br#"{"refund":{}}"#).unwrap();
        assert_eq!(deserialized, SomeMsg::Refund {});

        let expected = SomeMsg::ReleaseAll {
            image: "foo".to_string(),
            amount: 42,
            time: 18446744073709551615,
            karma: -17,
        };
        // &[u8]
        let deserialized: SomeMsg = from_json(
            br#"{"release_all":{"image":"foo","amount":42,"time":18446744073709551615,"karma":-17}}"#,
        )
        .unwrap();
        assert_eq!(deserialized, expected);

        // &str
        let deserialized: SomeMsg = from_json(
            r#"{"release_all":{"image":"foo","amount":42,"time":18446744073709551615,"karma":-17}}"#,
        )
        .unwrap();
        assert_eq!(deserialized, expected);
    }

    #[test]
    fn from_json_or_binary() {
        let msg = SomeMsg::Refund {};
        let serialized: Binary = to_json_binary(&msg).unwrap();

        let parse_binary: SomeMsg = from_json(&serialized).unwrap();
        assert_eq!(parse_binary, msg);

        let parse_slice: SomeMsg = from_json(serialized.as_slice()).unwrap();
        assert_eq!(parse_slice, msg);
    }

    #[test]
    fn to_json_vec_works_for_special_chars() {
        let msg = SomeMsg::Cowsay {
            text: "foo\"bar\\\"bla".to_string(),
        };
        let serialized = String::from_utf8(to_json_vec(&msg).unwrap()).unwrap();
        assert_eq!(serialized, r#"{"cowsay":{"text":"foo\"bar\\\"bla"}}"#);
    }

    #[test]
    fn from_json_works_for_special_chars() {
        let deserialized: SomeMsg = from_json(br#"{"cowsay":{"text":"foo\"bar\\\"bla"}}"#).unwrap();
        assert_eq!(
            deserialized,
            SomeMsg::Cowsay {
                text: "foo\"bar\\\"bla".to_string(),
            }
        );
    }

    #[test]
    fn to_json_string_works() {
        let msg = SomeMsg::Refund {};
        let serialized = to_json_string(&msg).unwrap();
        assert_eq!(serialized, r#"{"refund":{}}"#);

        let msg = SomeMsg::ReleaseAll {
            image: "foo".to_string(),
            amount: 42,
            time: 9007199254740999, // Number.MAX_SAFE_INTEGER + 7
            karma: -17,
        };
        let serialized = to_json_string(&msg).unwrap();
        assert_eq!(
            serialized,
            r#"{"release_all":{"image":"foo","amount":42,"time":9007199254740999,"karma":-17}}"#
        );
    }

    macro_rules! test_integer {
        ($($ty:ty),+$(,)?) => {
            $(
                ::paste::paste! {
                    #[property_test]
                    fn [<test_ $ty:snake:lower _encoding>](input: $ty) {
                        let primitive = input.get();

                        // Verify that the serialization is the same as the primitive
                        let serialized = to_json_string(&input).unwrap();
                        let serialized_primitive = to_json_string(&primitive).unwrap();
                        prop_assert_eq!(serialized.as_str(), serialized_primitive.as_str());

                        // Verify that the serialized primitive can be deserialized
                        let deserialized: $ty = from_json(serialized_primitive).unwrap();
                        assert_eq!(deserialized, input);

                        // Verify that zero is not allowed
                        assert!(from_json::<$ty>("0").is_err());

                        // Verify that the msgpack encoding is the same as the primitive
                        let serialized = to_msgpack_vec(&input).unwrap();
                        let serialized_primitive = to_msgpack_vec(&primitive).unwrap();
                        prop_assert_eq!(serialized.as_slice(), serialized_primitive.as_slice());

                        // Verify that the serialized primitive can be deserialized
                        let deserialized: $ty = from_msgpack(&serialized_primitive).unwrap();
                        prop_assert_eq!(deserialized, input);
                    }
                }
            )+
        };
    }

    test_integer! {
        NonZeroU8,
        NonZeroU16,
        NonZeroU32,
        NonZeroU64,
        NonZeroU128,
    }

    test_integer! {
        NonZeroI8,
        NonZeroI16,
        NonZeroI32,
        NonZeroI64,
        NonZeroI128,
    }
}
