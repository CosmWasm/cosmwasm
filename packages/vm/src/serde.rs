//! This file simply re-exports some methods from serde_json
//! The reason is two fold:
//! 1. To easily ensure that all calling libraries use the same version (minimize code size)
//! 2. To allow us to switch out to eg. serde-json-core more easily
use serde::{Deserialize, Serialize};
use std::any::type_name;

use crate::errors::{VmError, VmResult};
use crate::security_limits::{
    enforce_deserialization_limit, MAX_DESERIALIZATION_BYTES, MAX_DESERIALIZATION_DEPTH,
};

/// Deserializes JSON data into a document of type `T`.
///
/// The deserialization limit ensure it is not possible to slow down the execution by
/// providing overly large JSON documents.
pub fn from_slice<'a, T>(value: &'a [u8], deserialization_limit: usize) -> VmResult<T>
where
    T: Deserialize<'a>,
{
    // Enforce a global maximum regardless of what limit is passed in
    let enforced_limit = enforce_deserialization_limit(value.len(), deserialization_limit);

    if value.len() > enforced_limit {
        return Err(VmError::deserialization_limit_exceeded(
            value.len(),
            enforced_limit,
        ));
    }

    // Create a depth-limiting deserializer
    let mut deserializer = serde_json::Deserializer::from_slice(value);
    deserializer.disable_recursion_limit();
    let deserializer = serde_json::Deserializer::with_recursion_limit(
        deserializer,
        MAX_DESERIALIZATION_DEPTH as usize,
    );

    // Use the depth-limited deserializer
    let result = T::deserialize(deserializer);
    result.map_err(|e| VmError::parse_err(type_name::<T>(), e))
}

pub fn to_vec<T>(data: &T) -> VmResult<Vec<u8>>
where
    T: Serialize + ?Sized,
{
    serde_json::to_vec(data).map_err(|e| VmError::serialize_err(type_name::<T>(), e))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;

    const LIMIT: usize = 20_000;

    #[derive(Serialize, Deserialize, Debug, PartialEq)]
    #[serde(rename_all = "snake_case")]
    enum SomeMsg {
        Refund {},
        ReleaseAll {
            image: String,
            amount: u32,
            time: u64,
            karma: i32,
        },
        Cowsay {
            text: String,
        },
    }

    #[test]
    fn from_slice_works() {
        let deserialized: SomeMsg = from_slice(br#"{"refund":{}}"#, LIMIT).unwrap();
        assert_eq!(deserialized, SomeMsg::Refund {});

        let deserialized: SomeMsg = from_slice(
            br#"{"release_all":{"image":"foo","amount":42,"time":18446744073709551615,"karma":-17}}"#, LIMIT
        )
        .unwrap();
        assert_eq!(
            deserialized,
            SomeMsg::ReleaseAll {
                image: "foo".to_string(),
                amount: 42,
                time: 18446744073709551615,
                karma: -17
            }
        );
    }

    #[test]
    fn from_slice_works_for_special_chars() {
        let deserialized: SomeMsg =
            from_slice(br#"{"cowsay":{"text":"foo\"bar\\\"bla"}}"#, LIMIT).unwrap();
        assert_eq!(
            deserialized,
            SomeMsg::Cowsay {
                text: "foo\"bar\\\"bla".to_string(),
            }
        );
    }

    #[test]
    fn from_slice_errors_when_exceeding_deserialization_limit() {
        let result = from_slice::<SomeMsg>(br#"{"refund":{}}"#, 5);
        match result.unwrap_err() {
            VmError::DeserializationLimitExceeded {
                length, max_length, ..
            } => {
                assert_eq!(length, 13);
                assert_eq!(max_length, 5);
            }
            err => panic!("Unexpected error: {err:?}"),
        }
    }

    #[test]
    fn from_slice_enforces_global_limit() {
        // Test that even if a large limit is provided, it's capped by MAX_DESERIALIZATION_BYTES
        let very_large_limit = MAX_DESERIALIZATION_BYTES * 2;
        let data = vec![b'{'; MAX_DESERIALIZATION_BYTES + 1000]; // Data slightly larger than the max
        let result = from_slice::<serde_json::Value>(&data, very_large_limit);
        match result.unwrap_err() {
            VmError::DeserializationLimitExceeded {
                length, max_length, ..
            } => {
                assert_eq!(length, MAX_DESERIALIZATION_BYTES + 1000);
                assert_eq!(max_length, MAX_DESERIALIZATION_BYTES);
            }
            err => panic!("Unexpected error: {err:?}"),
        }
    }

    #[test]
    fn from_slice_prevents_deeply_nested_json() {
        // Create a deeply nested JSON structure beyond our recursion limit
        let mut json = String::from("{\"data\":");
        for _ in 0..MAX_DESERIALIZATION_DEPTH as usize + 5 {
            json.push_str("{\"data\":");
        }
        json.push_str("\"value\"");
        for _ in 0..MAX_DESERIALIZATION_DEPTH as usize + 5 {
            json.push('}');
        }

        // This should fail due to recursion depth
        let result = from_slice::<serde_json::Value>(json.as_bytes(), LIMIT);
        assert!(result.is_err());
        let err = result.unwrap_err();
        match err {
            VmError::ParseErr { .. } => {
                // We expect a parsing error related to recursion depth
                assert!(
                    err.to_string().contains("recursion")
                        || err.to_string().contains("depth")
                        || err.to_string().contains("nested")
                );
            }
            err => panic!("Unexpected error: {err:?}"),
        }
    }

    #[test]
    fn to_vec_works() {
        let msg = SomeMsg::Refund {};
        let serialized = to_vec(&msg).unwrap();
        assert_eq!(serialized, br#"{"refund":{}}"#);

        let msg = SomeMsg::ReleaseAll {
            image: "foo".to_string(),
            amount: 42,
            time: 9007199254740999, // Number.MAX_SAFE_INTEGER + 7
            karma: -17,
        };
        let serialized = String::from_utf8(to_vec(&msg).unwrap()).unwrap();
        assert_eq!(
            serialized,
            r#"{"release_all":{"image":"foo","amount":42,"time":9007199254740999,"karma":-17}}"#
        );
    }

    #[test]
    fn to_vec_works_for_special_chars() {
        let msg = SomeMsg::Cowsay {
            text: "foo\"bar\\\"bla".to_string(),
        };
        let serialized = String::from_utf8(to_vec(&msg).unwrap()).unwrap();
        assert_eq!(serialized, r#"{"cowsay":{"text":"foo\"bar\\\"bla"}}"#);
    }
}
