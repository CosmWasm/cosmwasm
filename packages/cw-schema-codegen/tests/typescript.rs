use arbitrary::Arbitrary;
use core::str;
use cw_schema::Schemaifier;
use serde::{Deserialize, Serialize};
use std::{
    fs::File,
    io::Write,
    process::{Command, Stdio},
};

#[derive(Arbitrary, Schemaifier, Debug, Deserialize, PartialEq, Serialize)]
struct Owo {
    field_1: u32,
    field_2: String,
}

#[derive(Arbitrary, Schemaifier, Debug, Deserialize, PartialEq, Serialize)]
struct Uwu(String, u32);

#[derive(Arbitrary, Schemaifier, Debug, Deserialize, PartialEq, Serialize)]
struct Òwó;

#[derive(Schemaifier, Debug, Deserialize, PartialEq, Serialize)]
pub enum Empty {}

#[derive(Arbitrary, Schemaifier, Debug, Deserialize, PartialEq, Serialize)]
enum Hehehe {
    A,
    B(u32),
    C { field: String },
}

#[derive(Debug, Deserialize, PartialEq, Serialize)]
#[serde(untagged)]
enum Combined {
    Owo(Owo),
    Uwu(Uwu),
    Òwó(Òwó),
    Empty(Empty),
    Hehehe(Hehehe),
}

macro_rules! impl_from {
    ($ty:ident) => {
        impl From<$ty> for Combined {
            fn from(ty: $ty) -> Combined {
                Combined::$ty(ty)
            }
        }
    };
}

impl_from!(Owo);
impl_from!(Uwu);
impl_from!(Òwó);
impl_from!(Empty);
impl_from!(Hehehe);

#[test]
fn codegen_snap() {
    // generate the schemas for each of the above types
    let schemas = [
        cw_schema::schema_of::<Owo>(),
        cw_schema::schema_of::<Uwu>(),
        cw_schema::schema_of::<Òwó>(),
        cw_schema::schema_of::<Empty>(),
        cw_schema::schema_of::<Hehehe>(),
    ];

    // run the codegen to typescript
    for schema in schemas {
        let cw_schema::Schema::V1(schema) = schema else {
            panic!();
        };

        let output = schema
            .definitions
            .iter()
            .map(|node| {
                let mut buf = Vec::new();
                cw_schema_codegen::typescript::process_node(&mut buf, &schema, node, true).unwrap();
                String::from_utf8(buf).unwrap()
            })
            .collect::<String>();

        insta::assert_snapshot!(output);
    }
}

fn wrap<T: for<'a> Arbitrary<'a> + Into<Combined>>(
    stuff: &mut arbitrary::Unstructured,
) -> Combined {
    T::arbitrary(stuff).unwrap().into()
}

fn type_name<T>() -> String {
    let name = std::any::type_name::<T>()
        .replace("::", "_")
        .replace(['<', '>'], "_");

    format!("{name}Schema")
}

#[test]
fn assert_validity() {
    #[allow(clippy::type_complexity)]
    let schemas: &[(_, fn(&mut arbitrary::Unstructured) -> Combined, _)] = &[
        (
            cw_schema::schema_of::<Owo>(),
            wrap::<Owo>,
            type_name::<Owo>(),
        ),
        (
            cw_schema::schema_of::<Uwu>(),
            wrap::<Uwu>,
            type_name::<Uwu>(),
        ),
        (
            cw_schema::schema_of::<Òwó>(),
            wrap::<Òwó>,
            type_name::<Òwó>(),
        ),
        // `Empty` is a non-constructable type
        /*(
            cw_schema::schema_of::<Empty>(),
            wrap::<Empty>,
            type_name::<Empty>(),
        ),*/
        (
            cw_schema::schema_of::<Hehehe>(),
            wrap::<Hehehe>,
            type_name::<Hehehe>(),
        ),
    ];

    let e2e_dir = format!("{}/tests/ts-e2e", env!("CARGO_MANIFEST_DIR"));
    let gen_file_path = format!("{}/src/gen.ts", e2e_dir);

    // make sure the dependencies are installed
    let install_status = Command::new("npm")
        .arg("i")
        .current_dir(&e2e_dir)
        .status()
        .unwrap();
    assert!(install_status.success());

    let random_data: [u8; 255] = rand::random();
    let mut unstructured = arbitrary::Unstructured::new(&random_data);
    for (schema, arbitrary_gen, type_name) in schemas {
        let cw_schema::Schema::V1(schema) = schema else {
            unreachable!();
        };

        let output = schema
            .definitions
            .iter()
            .map(|node| {
                let mut buf = Vec::new();
                cw_schema_codegen::typescript::process_node(&mut buf, schema, node, true).unwrap();
                String::from_utf8(buf).unwrap()
            })
            .collect::<String>();

        let mut gen_file = File::create(&gen_file_path).unwrap();
        gen_file.write_all(output.as_bytes()).unwrap();

        let data = arbitrary_gen(&mut unstructured);
        let serialized = serde_json::to_string(&data).unwrap();

        let mut child = Command::new("npm")
            .args(["test", type_name])
            .current_dir(&e2e_dir)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .spawn()
            .unwrap();

        {
            let mut stdin = child.stdin.take().unwrap();
            stdin.write_all(serialized.as_bytes()).unwrap();
        }

        let proc_output = child.wait_with_output().unwrap();
        assert!(
            proc_output.status.success(),
            "failed with object: {data:#?}; json: {serialized}; schema: {output}"
        );

        let stdout = str::from_utf8(&proc_output.stdout).unwrap();
        let stdout = stdout.lines().last().unwrap();
        let deserialized: Combined = serde_json::from_str(stdout).unwrap_or_else(|err| {
            panic!("{err:?}; input: {serialized}, output: {stdout}");
        });

        assert_eq!(data, deserialized);
    }
}
