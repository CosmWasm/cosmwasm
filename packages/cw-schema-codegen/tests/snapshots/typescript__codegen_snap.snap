---
source: packages/cw-schema-codegen/tests/typescript.rs
expression: output
---
// This code is @generated by cw-schema-codegen. Do not modify this manually.


import { z } from 'zod';


/**

 */

const typescript_OwoSchema = 

    z.object({

        /**
        
         */

        field_1: z.string().or(z.number()),

        /**
        
         */

        field_2: z.string(),

})

;

type typescript_Owo = z.infer<typeof typescript_OwoSchema>;

export { typescript_Owo, typescript_OwoSchema };
